internal/resources/resources.go

```diff
@@ -0,0 +1,130 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+type ResourceMeta struct {
+	Name      string
+	Namespace string
+}
+
+type Resource struct {
+	ResourceMeta
+	GVK schema.GroupVersionKind
+}
+
+type ResourceConfiguration struct {
+	GroupVersionKind schema.GroupVersionKind
+	ControlledList   []ResourceMeta
+}
+
+type resourceFinderConfiguration struct {
+	Resources []ResourceConfiguration
+}
+
+type ResourcesFinder struct {
+	ctx           context.Context
+	logger        logr.Logger
+	client        client.Client
+	configuration resourceFinderConfiguration
+}
+
+type resourceCondition func(context.Context, client.Client, Resource) bool
+
+var noMatchesForKind = regexp.MustCompile("no matches for kind")
+var couldNotFindReqResource = regexp.MustCompile("could not find the requested resource")
+
+var ReadFileHandle = os.ReadFile
+
+func NewResourcesFinderFromConfigYaml(ctx context.Context, client client.Client, logger logr.Logger, path string) (*ResourcesFinder, error) {
+	configYaml, err := ReadFileHandle(path)
+	if err != nil {
+		return nil, err
+	}
+	var finderConfiguration resourceFinderConfiguration
+	err = yaml.Unmarshal(configYaml, &finderConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, resource := range finderConfiguration.Resources {
+		for _, meta := range resource.ControlledList {
+			_, err := regexp.Compile(meta.Name)
+			if err != nil {
+				return nil, fmt.Errorf("configuration yaml regex check failed for \"%s\": %s", meta.Name, err)
+			}
+
+			_, err = regexp.Compile(meta.Namespace)
+			if err != nil {
+				return nil, fmt.Errorf("configuration yaml regex check failed for \"%s\": %s", meta.Namespace, err)
+			}
+		}
+	}
+
+	return &ResourcesFinder{
+		ctx:           ctx,
+		logger:        logger,
+		client:        client,
+		configuration: finderConfiguration,
+	}, nil
+}
+
+func (i *ResourcesFinder) FindUserCreatedResources(checkIf resourceCondition) ([]Resource, error) {
```

Comment:
Some documentation, especially how the provided function is used might be helpful. `checkIf` seems also very generic, maybe we can find a better name or make it a "boolean name" like `isResourceRelevant`.

Reply:
The goal of it was to be generic condition. I do not care much for names, will rename it to `isResourceRelevant`. If you ask me to add comments to the code, I politely refuse 😜